File: write-tests/templates/test_template.py
"""
Test Template for Python (pytest)

Usage:
    pytest tests/test_<module>.py -v
    pytest tests/test_<module>.py::TestClassName::test_method -v
"""
import pytest
from unittest.mock import Mock, patch, MagicMock
from datetime import datetime

# Import your module
# from myapp.module import MyClass, my_function


class TestMyFunction:
    """Tests for my_function"""

    def test_basic_usage(self):
        """Test normal expected behavior"""
        # Arrange
        input_data = "test"
        expected = "expected_output"

        # Act
        result = my_function(input_data)

        # Assert
        assert result == expected

    def test_edge_case_empty_input(self):
        """Test with empty input"""
        result = my_function("")
        assert result is None  # or whatever expected

    def test_edge_case_none_input(self):
        """Test with None input"""
        with pytest.raises(ValueError):
            my_function(None)

    def test_large_input(self):
        """Test with large input"""
        large_input = "x" * 10000
        result = my_function(large_input)
        assert len(result) <= 10000

    @pytest.mark.parametrize("input_val,expected", [
        ("a", "A"),
        ("hello", "HELLO"),
        ("123", "123"),
    ])
    def test_multiple_inputs(self, input_val, expected):
        """Test multiple input/output combinations"""
        assert my_function(input_val) == expected


class TestMyClass:
    """Tests for MyClass"""

    @pytest.fixture
    def instance(self):
        """Create a fresh instance for each test"""
        return MyClass(name="test")

    @pytest.fixture
    def mock_dependency(self):
        """Mock external dependency"""
        with patch("myapp.module.external_service") as mock:
            mock.return_value = {"status": "ok"}
            yield mock

    def test_initialization(self, instance):
        """Test class initialization"""
        assert instance.name == "test"
        assert instance.created_at is not None

    def test_method_success(self, instance, mock_dependency):
        """Test method with mocked dependency"""
        result = instance.do_something()
        
        assert result["success"] is True
        mock_dependency.assert_called_once()

    def test_method_failure(self, instance, mock_dependency):
        """Test method when dependency fails"""
        mock_dependency.side_effect = ConnectionError("Network error")
        
        with pytest.raises(ConnectionError):
            instance.do_something()


class TestIntegration:
    """Integration tests"""

    @pytest.fixture
    def db_session(self):
        """Setup test database"""
        # Setup
        session = create_test_session()
        yield session
        # Teardown
        session.rollback()
        session.close()

    @pytest.mark.integration
    def test_full_workflow(self, db_session):
        """Test complete user workflow"""
        # Create
        user = User(name="test")
        db_session.add(user)
        db_session.commit()

        # Read
        found = db_session.query(User).filter_by(name="test").first()
        assert found is not None
        assert found.name == "test"


# Async tests (pytest-asyncio)
class TestAsyncFunctions:
    """Tests for async functions"""

    @pytest.mark.asyncio
    async def test_async_function(self):
        """Test async function"""
        result = await async_fetch_data()
        assert result is not None


# Fixtures for common test data
@pytest.fixture
def sample_user():
    return {
        "id": 1,
        "name": "Test User",
        "email": "test@example.com"
    }


@pytest.fixture
def sample_config():
    return {
        "debug": True,
        "database_url": "sqlite:///:memory:"
    }

File: write-tests/templates/test_template.ts
/**
 * Test Template for TypeScript (Vitest/Jest)
 * 
 * Usage:
 *   npx vitest run src/__tests__/module.test.ts
 *   npx jest src/__tests__/module.test.ts
 */
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
// For Jest: import { describe, it, expect, beforeEach, afterEach, jest } from '@jest/globals';

// Import your module
// import { myFunction, MyClass } from '../module';

describe('myFunction', () => {
  it('should return expected output for normal input', () => {
    // Arrange
    const input = 'test';
    const expected = 'expected';

    // Act
    const result = myFunction(input);

    // Assert
    expect(result).toBe(expected);
  });

  it('should handle empty input', () => {
    expect(myFunction('')).toBeNull();
  });

  it('should throw on null input', () => {
    expect(() => myFunction(null)).toThrow('Input required');
  });

  it.each([
    ['a', 'A'],
    ['hello', 'HELLO'],
    ['123', '123'],
  ])('should transform "%s" to "%s"', (input, expected) => {
    expect(myFunction(input)).toBe(expected);
  });
});

describe('MyClass', () => {
  let instance: MyClass;

  beforeEach(() => {
    instance = new MyClass({ name: 'test' });
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('should initialize with correct values', () => {
    expect(instance.name).toBe('test');
    expect(instance.createdAt).toBeInstanceOf(Date);
  });

  it('should process data correctly', async () => {
    const result = await instance.process({ value: 42 });
    
    expect(result).toMatchObject({
      success: true,
      value: 42,
    });
  });

  describe('with mocked dependency', () => {
    beforeEach(() => {
      vi.spyOn(externalService, 'fetch').mockResolvedValue({ status: 'ok' });
    });

    it('should call external service', async () => {
      await instance.doSomething();
      
      expect(externalService.fetch).toHaveBeenCalledOnce();
      expect(externalService.fetch).toHaveBeenCalledWith(
        expect.objectContaining({ id: instance.id })
      );
    });

    it('should handle service failure', async () => {
      vi.spyOn(externalService, 'fetch').mockRejectedValue(new Error('Network error'));
      
      await expect(instance.doSomething()).rejects.toThrow('Network error');
    });
  });
});

describe('API Integration', () => {
  // Skip in CI if needed
  it.skipIf(process.env.CI)('should connect to real API', async () => {
    const response = await fetch('https://api.example.com/health');
    expect(response.ok).toBe(true);
  });
});

// Snapshot testing
describe('Component rendering', () => {
  it('should match snapshot', () => {
    const output = renderToString(<MyComponent name="test" />);
    expect(output).toMatchSnapshot();
  });
});

// Test utilities
const createTestUser = (overrides = {}) => ({
  id: 1,
  name: 'Test User',
  email: 'test@example.com',
  ...overrides,
});

const createTestConfig = (overrides = {}) => ({
  debug: true,
  apiUrl: 'http://localhost:3000',
  ...overrides,
});

// Mock factory
const createMockRepository = () => ({
  findById: vi.fn(),
  save: vi.fn(),
  delete: vi.fn(),
});
